[요구사항 7] 하리 제출합니다 🍉

LAH1203 opened the pull request:
오홍홍 이번 요구사항엔 음료가 추가되었네요 !

- [x] 다른 가격의 콜라와 사이다가 추가된다.

저는 가격이 다르다는 게 콜라와 사이다가 다르다는 건지 아니면 콜라 여러 종류 사이의 가격과 사이다 여러 종류 사이의 가격이 다르다는 건지 헷갈려서 그냥 후자로 보기로 했습니다!
그래서 콜라(코카콜라, 펩시), 사이다(칠성사이다, 스프라이트) 총 4가지의 음료가 추가되었습니다 🔥

### [🌼 데모 페이지 🌼](https://lah1203.github.io/2022-looking-for-the-sound-of-an-object/)

<br />

---

<br />

사실 이번에는 단순히 음료의 종류만 추가되는 것이라 새롭게 컴포넌트의 소리를 들은 건 없습니다. 다만 지난번에 구현을 너무 급하게 한 터라 미처 신경 쓰지 못했던 컴포넌트들의 소리를 들어보았어요 👂

<br />

### 🗣 첫 번째 요구사항

> 나는 딱 한 컴포넌트에서만 쓰이는 하위 컴포넌트인데 어째서 `components`라는 공용 폴더에서 관리하는거야? 위치 조정 해줘!

그래서 `ChangeInput`, `ChangeOutput`은 `Change` 폴더 내로, `MenuButton`은 `Menu` 폴더 내로 옮겨주었어요.

<br />

### 🗣 두 번째 요구사항

> 나는 사용하지 않는 value나 setter를 가지고 있지 않고 싶어! 방법이 없을까?

찾아보니 Recoil 자체에서 지원하는 것에 `useRecoilState` 외에도 `useRecoilValue`, `useSetRecoilState`가 있더라고요. 바로 바꿔주었습니다 🙌
그렇게 해주니 불필요한 리렌더링이 많이 사라졌네요 🥳

<br />

### 🗣 세 번째 요구사항

> 어차피 잔돈 충전 input을 uncontrolled로 사용할거라면 잔돈 충전 성공할 때마다 input을 비워주고 포커싱해주면 안될까?

ㅇㅋ

<br />

### 🗣 네 번째 요구사항

> 음료 제조 중에도 잔돈 반환이 가능한 건 메시지도 겹치고 조금 불편한 것 같아. 이 점도 개선 부탁!

(아이코 이 부분을 안해줬었네,, 미안타 친구야)
음료 제조 중에 반환 버튼 클릭 시 메뉴 버튼과 동일하게 alert으로 막아주었어요.

<br />

---

<br />

우선은 이 정도만 개선해보았습니다 🌼
리뷰 잘 부탁드립니다 :) 요즘 `UI / UX`에 조금 관심이 생겨서 그 부분에 대한 피드백도 환영입니다 !

LAH1203 committed:
docs: 7단계 요구사항 작성

LAH1203 committed:
docs: README 업데이트

LAH1203 committed:
feat: 각각 다른 가격의 콜라, 사이다 추가

LAH1203 committed:
fix: disabled props 추가에 따라 불필요한 예외 처리 사항 제거

LAH1203 committed:
style: 메뉴 버튼 스타일링 변경

LAH1203 committed:
style: 디스펜서 스타일링 변경

LAH1203 committed:
style: 가이드 메시지 스타일링 변경

LAH1203 committed:
style: 메뉴 버튼 hover 애니메이션 추가

LAH1203 committed:
style: components 폴더 위치 정리

- 관련 있는 하위 폴더는 상위 폴더 내에 위치하도록 변경

LAH1203 committed:
fix: 초기 가이드 메시지 변경

LAH1203 committed:
refactor: 사용하지 않는 stateValue, setter 제거

- useRecoilState 대신 useRecoilValue, useSetRecoilState 적용

LAH1203 committed:
refactor: 음료 제조 과정을 함수로 분리

LAH1203 committed:
refactor: as를 없애고 타입 가드로 변경

LAH1203 committed:
feat: 잔돈 충전 input에 ref 추가

- 잔돈 충전 성공 시 input value 비우기
- 잔돈 충전 성공 시 input에 focus

LAH1203 committed:
feat: 음료 제조 중에 반환 버튼 클릭시 막는 기능 추가

LAH1203 committed:
feat: 잔돈 충전 input에 min, max 추가

soyi47 commented on src/styles/theme.ts:
```diff
@@ -2,11 +2,12 @@ const theme = {
-    brown001: '#946d59',
-    red001: '#ff1100',
-    green001: '#04ff00',
+    gray002: '#E0E0E0',
+    white001: '#FFFFFF',
+    brown001: '#94716B',
+    brown002: '#B79891',
+    red001: '#FF1100',
+    green001: '#04FF00',
```
대문자 컨벤션 정리 좋아요~ 👍

soyi47 commented on src/components/Change/ChangeInput/index.tsx:
```diff
@@ -0,0 +1,55 @@
+function ChangeInput() {
+  const setGuideMessage = useSetRecoilState(guideState);
+  const setChange = useSetRecoilState(changeState);
+  const changeRef = useRef<HTMLInputElement>(null);
+
+  const chargeChange = (e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (
+      !(e.target instanceof HTMLFormElement) ||
+      !(e.target.elements[0] instanceof HTMLInputElement)
+    )
+      return;
+
+    const money = e.target.elements[0].valueAsNumber;
+
+    try {
+      validateChangeInput(money);
+    } catch ({ message }) {
+      alert(message);
+      return;
+    }
+
+    setChange(prevState => prevState + money);
+
+    setGuideMessage('투입되었습니다.');
+
+    if (!changeRef.current) return;
+
+    changeRef.current.value = '';
+  };
```
비제어 컴포넌트라 ref를 이용하여 input을 clear 하나보네요! 
ref가 없어도 `const money = e.target.elements[0].valueAsNumber;` 라인과 같이 `form.elements`로 접근해서 input을 clear할 수도 있을 것 같아요.
리코일로 관리하는 change 상태를 이용하여 제어 컴포넌트로 관리할 수도 있을 것 같은데, 혹시 비제어로 관리한 이유가 있나요?

LAH1203 replied:
우선 저는 폼 태그를 쓰면 제어보단 비제어로 하는 것을 좀 더 선호하는 편입니다. 특별한 이유는 없는데, `useState`로 관리하는 상태가 많아진다면..? 을 생각하니 자연스레 비제어로 해도 될 것들은 상태로 관리하지 않게 되더라고요. 여기에서도 비제어로 관리한 것은 습관입니다 ㅎㅎ

소피아가 말씀해주신대로 form의 엘리먼트로 접근해서 리셋해도 될 것 같네요! 사실 저 부분은 포커싱까지 ref로 썼던 터라 저렇게 되었는데, (나중에 포커싱은 HTML의 autofocus를 사용하는 것으로 바꿨습니다) ref를 빼는 부분은 따로 고려하지 않았네용 👀

피드백 반영하여 수정하였습니다 ❤️ [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23/commits/8b14d6e29bb18190f896930240f518f628fd5f7c)

soyi47 commented on src/components/Change/ChangeInput/index.tsx:
```diff
@@ -0,0 +1,55 @@
+  };
+
+  return (
+    <S.ChangeInputForm onSubmit={chargeChange}>
```
submit 이벤트 핸들러 함수로 이름이 조금 아쉬운 것 같아요! 
해당 함수가 잔돈을 충전하는 것 뿐만 아니라, 가이드 메세지를 업데이트하고, 또 input을 clear하는 역할도 하고 있다는 점을 고려해서 다시 이름을 고민해볼 수 있을 것 같아요~

조금 다른 이야긴데 charge가 동사로는 충전하다, 명사로는 요금이라는 의미가 있어서 저는 네이밍 할 때 혼란스럽더라고요. T.T 좋은 표현이 없을까 계속 고민 중인데, 하리 생각은 어때요~? 좋은 아이디어 있으면 공유해주세용 🥰

LAH1203 replied:
흠 항상 `한 함수가 한 가지 일만 하게 만들어라`가 참 어려운 것 같아요,,
저는 주로 `on{Event}{Element}` 이런 식으로 네이밍을 하곤 했는데 전에 이러면 이 함수가 무슨 일을 하는지 명확하게 알지 못한다는 얘기를 듣고 이젠 그 함수가 무슨 일을 하는지를 이름에 담아보려고 이렇게 지어봤는데요. 말씀해주신 것처럼 이 함수는 1) 입력된 값이 맞는지 검증하고, 2) 잔돈을 충전하고, 3) 가이드 메시지를 업데이트하고, 4) input을 clear하는 일까지 하고 있으니, `chargeChange`라는 이름은 일부 행동만 지칭하는 말이라 맞지 않는 것 같네요 💦
저는 `charge`가 요금과 헷갈려 혼란스럽기보다는 뒤에 잔돈이라는 뜻의 `change`와 붙여서 사용하게 되니 둘이 비슷하게 생겨서 그 부분이 헷갈리더라고요 🥹 마찬가지로 잔돈을 충전한다는 의미의 다른 이름이 없을까 고민하고 있었습니다,, 근데 지금 생각해보니 잔돈을 충전하기도 하지만 투입한다고도 볼 수 있을 것 같네요 ! 실제로 가이드 메시지에도 '투입되었습니다.'라고 표기해주고 있고요. 그럼 투입한다는 의미에서 `inject`를 쓸 수도 있을 것 같아요 🥰
우선 해당 함수는 잔돈 투입을 시도하고 있으니 `tryInjectChange`로 하고, 잔돈 투입 성공 로직을 `successInject`로 분리하여 사용해주도록 하겠습니당 🙌

[commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23/commits/ce2f5c8701f1613715551a652f11850f25273f25)

soyi47 replied:
투입이라는 표현을 사용하는 것도 좋네요! 
저는 이전보다 지금 try, success로 나누어준 게 더 깔끔한 것 같아요! 좋습니다~ 👍

soyi47 commented on src/components/Dispenser/index.tsx:
```diff
@@ -1,12 +1,12 @@
-import { useRecoilState } from 'recoil';
+import { useRecoilValue } from 'recoil';
 import { dispenserState } from 'recoil/states';
 
 import { DISPENSER_STATE } from 'constants/dispenser';
 
 import * as S from './index.styled';
 
 function Dispenser() {
-  const [dispenser, setDispenser] = useRecoilState(dispenserState);
+  const dispenser = useRecoilValue(dispenserState);
```
하리 덕분에 리코일 코드도 접해보네요 🧐

soyi47 commented on src/components/Change/ChangeInput/index.tsx:
```diff
@@ -0,0 +1,55 @@
+  return (
+    <S.ChangeInputForm onSubmit={chargeChange}>
+      <input
+        type="number"
+        min="500"
+        max="999999500"
+        ref={changeRef}
+        autoFocus
```
step 속성도 같이 주면 더 좋을 것 같아요~

LAH1203 replied:
앗 그러네요! 지금은 500원 단위로 입력해달라고 요청하고 있으니, step을 500으로 주도록 할게요 :)
알려주셔서 감사합니당 😘

[commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23/commits/3254b4a3fc00bfabe580dd46d7d3b4a48e5950ae)

soyi47 replied:
수정하면서 상수화도 챙겼군요! 좋아좋아~ 👍

soyi47 commented on src/components/Change/ChangeOutput/index.tsx:
```diff
@@ -1,14 +1,22 @@
-  const [guideMessage, setGuideMessage] = useRecoilState(guideState);
+  const setGuideMessage = useSetRecoilState(guideState);
   const [change, setChange] = useRecoilState(changeState);
+  const dispenser = useRecoilValue(dispenserState);
 
   const returnChange = () => {
+    if (dispenser === DISPENSER_STATE.MAKING) {
+      alert('음료 제조가 끝난 후 반환해주세요.');
+      return;
+    }
+
```
dispenser 상태를 리코일로 관리하는 것 같네요. 
그렇다면,,, `dispenser === DISPENSER_STATE.MAKING`을 반환 버튼의 disabled 값으로 주는 건 어떤가요?! 
지금처럼 returnCharge 함수에서 dispenser 상태에 따른 방어 코드가 들어가는 건 아주 좋아요! 
더해서 반환 버튼에 disabled까지 줘서 사용자가 현재 반환 버튼을 누르지 못 한다는 걸 바로 알 수 있으면 더 좋을 것 같아요~

LAH1203 replied:
저도 원래는 할 수 없다는 의미로 버튼에 `disabled`를 주는 것을 좋아했었는데요. 우연히 예전에 들었던 수업자료를 보다가 '좋은 UX는 사용자에게 자유를 주어야 한다'는 누군가(정확히 기억이 안나요,,)의 주장을 보았습니다. 저도 그 의견에 어느 정도 공감했고요. 그래서 여기에서도 무작정 '사용할 수 없어!'하고 막는 것보단 일단 할 수 있도록 둔 다음 상황에 따라 경고를 주거나 처리하는 방식을 적용해보았습니다.
혹시 이 경우 어느 쪽이 사용자의 입장에서 당황스럽지 않고 좀 더 자연스럽다고 느껴질까요? 소피아의 의견도 들어보고 싶어요~!

soyi47 replied:
흐음,,, 어렵네요...! 
disabled를 줬을 때 좋은 점이 뭘까, 생각해보면
1. 자판기가 다른 동작을 처리 중일 때(음료를 만들고 있을 때)에는 disabled를 주는 걸, 사용자에게 자판기가 다른 동작을 처리 중임을 시각적으로 제공해준다는 관점에서 보면, 사용자에게 직관적인 정보를 준다는 점에서 장점이 있다는 생각이 들고
2. 개인적으로는 사용자가 어떤 동작을 언제 할지 모르는 상황에서, 기능을 열어뒀을 때 어떤 오류가 생길지 모르기 때문에 최대한 방어적으로 막아 놓고 싶은 것 같아요.

하지만 사용자에게 자유를 주어야 한다는 주장에도 공감은 됩니다! (정확한 근거나 예시 상황은 잘 떠오르지 않지만...)
고민해보고, 수정이 필요하면 다음 요구사항에 반영해보아요! 👍

soyi47 commented on src/components/GuideMessage/index.tsx:
```diff
@@ -1,12 +1,16 @@
 
-  return <S.GuideMessage>{guideMessage}</S.GuideMessage>;
+  return (
+    <S.GuideMessage>
+      <span>{guideMessage}</span>
+    </S.GuideMessage>
```
내부에서 `<span>`으로 감싸는 이유가 있나요~?

아 ㅋㅋㅋㅋ 묘하게 글자 밝기 바뀌는 이 애니메이션을 추가한 것 같은데, 그럼 스타일드 컴포넌트로 만든 아이들의 네이밍을 다시 생각해봐도 좋을 것 같아요~ ㅋㅋㅋㅋㅋ 귀여웡 🥰

LAH1203 replied:
ㅋㅋㅋㅋㅋㅋ애니메이션 때문인걸 바로 들켜버렸네용 🤣
네이밍은 한 번 바꿔보도록 할게요 ❤️

[commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23/commits/aafe5ab32d1ac6a0d569e0723985b3674454472a)

soyi47 reviewed (approved):
하리~ 못 본지 너무 오래 되었어~ 보고 싶네요 ㅠㅅㅠ 

하리와 컴포넌트의 대화 잘 봤습니다. 너무 좋아요. 😆
하리 자판기도 장난 없네요. 다들 아주 짱판기야...
하리 레포 받아서 직접 실행도 해봤답니다. 🧐
이번에 TS로 옮겨보려는 중인데 하리 코드 리뷰한 게 많이 도움될 것 같아요!
덕분에 리코일 코드도 접해보고, TS도 훔쳐보고 갑니다.

요구사항 관련해서, 저는 기존 음료와 새로 추가되는 콜라, 사이다 서로 가격이 다르면 된다고 생각했어요. 
그래서 콜라랑 사이다만 추가할 생각이랍니다.. 하리는 총 4개나 추가해줬네요. 👍

추가로 디스펜서 까만 네모 상자 안에 무언가 컵이나 음료가 나타날 것 같은데 나타나지 않는 게 조금 아쉬운 느낌이에요. 
나중에 멋지게 구현해보려고 준비 중인 걸까요~?

코멘트가 많지는 않아요. 
UI/UX 관련 피드백도 환영이래서 관련해서는 한 두개 정도 남겨봤습니다. 
다음 요구사항에 반영해도 되니 읽어보고 하리 의견 달아주세요! 🥰

LAH1203 committed:
refactor: 잔돈 투입 시 input 초기화를 form element에 접근해서 하는 방식으로 변경

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23#discussion_r905929452
- ref 제거

LAH1203 committed:
refactor: 잔돈 충전 함수 네이밍 변경 및 성공 로직 분리

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23#discussion_r905932538
- chargeChange → tryInjectChange
- successToInject

LAH1203 committed:
feat: 잔돈 입력 input step 부여

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23#discussion_r905936597
- 잔돈 규칙(min, max, unit) 상수화

LAH1203 committed:
refactor: 가이드 메시지 styled 네이밍 변경

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/23#discussion_r905940724

soyi47 commented:
하리~ 코멘트 잘 확인했습니다! 머지할게요!
벌써 방학 마지막 날이네요. 😥
오늘 하루 잘 보내고 내일 선릉에서 봅시다. 💕
